Add TypeScript definitions

Codesleuth opened the pull request:
This PR adds a bundled type definitions file to allow consumers using TypeScript to use the package without having to manually declare the main export as `any` or take extra effort to define the types themselves.

The benefit of this is also one of developer tooling, where users of editors which integrate the TypeScript server (such as VSCode) will see hints such as this:

![image](https://user-images.githubusercontent.com/5011956/36077064-c420ae02-0f5d-11e8-8a54-a75972c28de3.png)

If you disagree with directly bundling this definition file, I can close this PR and create a new [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped) respository that mirrors this one for developers who wish to install the types explicitly with `npm install @types/postgrator`.

It's up to you if you are OK with the typed being bundled here - the drawback of this is that the types definition file _must_ be updated to reflect the JavaScript code, each time it is changed. I'll gladly help with this, but this does take time and effort. The alternative suggestion has the same issue, but it's generally more accepted that outdated types are monkey-patched when required.

rickbergfalk commented:
Thanks so much @Codesleuth! Especially for such a nicely documented PR 👌  

I think I'm okay with this... I use vscode myself and have been interested in getting more formal and strict with the API interfaces. The code hints are amazing.

Is there any way this can be tested or checked prior to npm publishes or is it strictly a manual effort to keep updated?

Codesleuth commented:
Thanks for the quick feedback!

The best way to test this is to write integration tests which combine both the features of the package with the type definitions - these can be isolated at the integration point, but they must be written in TypeScript themselves. This is where it starts to get a bit intrusive, and where DefinitelyTyped definitions seem more appropriate (although they would have the least confidence).

I will add some tests which demonstrate this, and we can go from there. If you feel they go beyond the scope of the package, we can revisit it and see what compromise we can make.

rickbergfalk commented:
Some example tests would be great! I'm not too concerned about adding it to the package and would be happy to add this if there were some tests I could build off of.

Codesleuth commented:
I've been busy this week so far, but will be pushing something tomorrow for you to take a look at. Bear with me :feet: !

rickbergfalk commented:
No worries! Whenever works for you

Codesleuth commented:
@rickbergfalk I've added a cloned test of the `api.js` tests, in TypeScript, and updated the mocha test command to run it. You can see it in the output as `TypeScript:API`.

Ideally we could move every test file to TypeScript, and we should see all the tests still passing - that would give me 100% confidence at least.

Let me know your thoughts!

rickbergfalk commented:
Sorry for the delay in getting to this! I hope to take a look this weekend

rickbergfalk commented:
Alright I looked into it. I compared the api.js/api.ts tests and I was surprised at how few changes there were. I'm assuming they'd all be minimal changes like declaring types and not much else?

Definitely in support of bringing this in. 

Should I merge in and publish as is or would you prefer tests be converted entirely prior to that?

Codesleuth commented:
Sorry for the delay; hectic week!

I was thinking about what you said about it being very few changes, and it was a good point! I've pushed a change which adds more _left_ to _right_ type assertions (assignment of types) to harden the type checking. I've also converted the tests with promise chains to `async/await` for readability.

I hope it looks good to you as it is now. I don't plan on making any more changes, so merge at you discretion :+1:

Codesleuth commented:
Sorry, I know I said I wouldn't make more changes, but I was looking at the commit history and noticed a PR that was merged since opening this one. I've added a commit which adds the feature these types.

rickbergfalk commented:
Thanks Dave! Sorry for the delay in getting back to you. I've been in the middle of a move. Will take a look at the rest of this today

rickbergfalk commented:
Looks great! Thanks again!
